JKISM 2024 [DEV]

Tom-McQuillan opened the pull request:
## New Features and Bug Fixes
- #21 New Support VI -> "Select State(s)" allows conditional (Boolean or error) enqueuing of states
- #11 JKI SM Template  -> Error Handler is now appends the  previous state string (and raw arguments) to the error message
- #30 JKI SM Template  -> Added UI: Cleanup state (called by Macro: Exit)
- #29 JKI SM Template ->  Clears warnings in the Idle state
- #17 JKI SM Template ->  fixed typo in UI: Initialize state comment

## Developer Fixes
- Updated VIPC ->  include all package dependencies required to open the source and build VIs
- Updated VIPB -> Added new support VI to palate
- Updated VIPB -> Updated documentation

Tom-McQuillan committed:
Updated .vipc to include OpenG String Library

Tom-McQuillan committed:
FIX: #17 Updated comment in UI: Initalize

Tom-McQuillan committed:
Fix: #30 Added UI: Cleanup state to template

Tom-McQuillan committed:
Fix: #28 Added Select State(s) for String and Error (+Poly VI)

Tom-McQuillan committed:
Created Append Error Details to add State and Argument data to error message. #11

Tom-McQuillan committed:
Added Clear Warnings function. Added to Idle State #29

Tom-McQuillan committed:
Updated VIPB and copywrite notices to "2024" (from "2020")

Tom-McQuillan committed:
added openG Numeric to VIPC

Tom-McQuillan committed:
Added openg File and Variant configuration to VIPC

jimkring commented on source/State Machine Support API/Clear Warning.vi:
One way to simplify things would be to move the clearing of warnings into `Parse State Queue.vi`

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/99663b14-3bb0-41e0-8b8c-36c9c1b50575)

I'm curious what @philippe-guerit would think about this idea.

jimkring commented on source/State Machine Support API/Clear Warning.vi:
Clear Warning should probably be set to Subroutine Priority, since Parse State Queue (and others probably) are set to that priority.

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/89320aa3-df6f-4287-b1af-308114bd3509)

Note: We should add a Caraya test that verifies Execution settings (reentrancy, subroutine, etc.) by putting the VI Under Test in a static refnum and reading a property node to assert the correct settings.
- [x] Add Caraya test to verify execution settings of utility VIs

jimkring commented on source/State Machine Support API/Clear Warning.vi:
It might make sense to change to a smaller conpane pattern
![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/4b1a0fbe-cb27-454f-8be8-f51bf2f72225)

jimkring replied:
N/A, since this utility VI was removed

jimkring commented on source/State Machine Support API/Clear Warning.vi:
Probably want to comment the code (and wire the error into the select terminal)

> To clear "warnings" (i.e. when status is False, but code and source have values), we need to ensure that code and source are forced to defaults when there is no error.

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/0cf35f8f-7c12-423b-b4e0-b7fb1f5b2215)

Tom-McQuillan replied:
This function is no longer used

jimkring commented on source/State Machine Support API/Clear Warning.vi:
error in should probably be a required input
![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/9a201554-ec6b-4d6d-8c6a-2bac3498d736)

jimkring replied:
N/A

jimkring commented on source/State Machine Support API/Conditional State String.vi:
The various support VIs use gray/black blocks for the queue elements and green triangles for arrows.
- gray block signifies not yet queued
- black block signifies queued 

Would it make sense to update Conditional State String and Periodic State String to use consistent iconography? For example...

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/acc68154-5533-44c1-8a72-9181cb122a86)

jimkring replied:
@Tom-McQuillan and I (@jimkring) reviewed icons and thought these (below) were more consistent with the visual language already being used.

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/a3e6b0d2-c509-4f84-98e6-ab1f61331a3e)

Tom-McQuillan replied:
Done

Tom-McQuillan committed:
Updated Parse State Queue to capture warnings or error

Tom-McQuillan committed:
Added Test Warning VI (for review by @jimkring) Updated Append Error Details implementation

Tom-McQuillan committed:
Added comment in Error Handling state indicating Errors and Warnings are handled

jimkring reviewed (changes requested):
@Tom-McQuillan

Can you update the icons to look like the following?

![image](https://github.com/JKISoftware/JKI-State-Machine/assets/381432/4893d77e-9145-46ff-9064-abba581282a0)

jimkring reviewed (changes requested):
@Tom-McQuillan let's remove the Clear Warnings utility VI

Tom-McQuillan committed:
Created new Parse State Queue_2024 function to allow option upgrades to existing JKISMs. This change caused multiple files to update.

Tom-McQuillan committed:
Added unit tests for new Select State(s) function

Tom-McQuillan committed:
Added unit tests for Append Error Information and Parse State Queue 2024

Tom-McQuillan committed:
Fixed 2018 version of unit test to exclude previous state arguments for error handing

Tom-McQuillan committed:
Added unit tests for Warnings in Parse State Queue 2024

Tom-McQuillan committed:
Saved JKI SM Template with Idle, Timeout case visible

Tom-McQuillan committed:
Create testSavedWithIdleTimeoutVisible.vi

Tom-McQuillan committed:
Added testSavedWithIdleTimeoutVisible.vi to test case class

Since sourcecode is combined with the compiled code, all VIs in the class got updated.

jimkring commented on source/State Machine Templates/String-Based Queued State Machine (Basic 1.0).vi:
>  [616a596](https://github.com/JKISoftware/JKI-State-Machine/commit/616a596e298517a8ea3783dcccf51addc025ed88) Saved JKI SM Template with Idle, Timeout case visible

@Tom-McQuillan we could probably create a unit test to verify that the JKI SM Template has the `Idle, Timeout` case visible.

Tom-McQuillan replied:
Done
![image](https://github.com/JKISoftware/JKI-State-Machine/assets/104849466/d8bcbd10-4a5d-4f55-875c-005d7aacc293)

Tom-McQuillan commented:
> @Tom-McQuillan
> 
> Can you update the icons to look like the following?
> 
> ![image](https://private-user-images.githubusercontent.com/381432/346168904-4893d77e-9145-46ff-9064-abba581282a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2MzM4MjYsIm5iZiI6MTcyMDYzMzUyNiwicGF0aCI6Ii8zODE0MzIvMzQ2MTY4OTA0LTQ4OTNkNzdlLTkxNDUtNDZmZi05MDY0LWFiYmE1ODEyODJhMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQxNzQ1MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZDYzZDMyMGZjMjBlNWRhMDJmOWMwYTE4Y2I1Y2IxMGZjZGM1ZDE2Y2IzOWE5NWJhN2MyMjJkZmZmNjYyZWFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.zbyt6wR0kyKAC6xSu4sU9tgyp-eXdRRM8tfzRGyGXS4)

Done

Tom-McQuillan commented:
> @Tom-McQuillan let's remove the Clear Warnings utility VI

Done

Tom-McQuillan committed:
Updated vipb release description

jimkring committed:
upgraded sources to LV2020

- updated .vipc
- updated .vipb
- lvproj set to source only
- marked all project VIs as source only
- vendored the OpenG Periodic Trigger vi to remove dependency on oglib_time

jimkring committed:
added jki_labs_tool_vi_tester to .dragon dependencies

Tom-McQuillan committed:
Merge pull request #37 from JKISoftware/sources-upgrade-to-2020

Sources upgrade to 2020

Tom-McQuillan committed:
Create RELEASENOTES.md

First pass

Tom-McQuillan committed:
Moved release notes on disk and added content

Tom-McQuillan committed:
Added separated source code unit test

jimkring committed:
added a what's new in 2024 document that's more long form.

Tom-McQuillan committed:
Merge pull request #40 from JKISoftware/Seperate-Source-Code-unit-tests

Separate source code unit tests

Tom-McQuillan committed:
#43 Remove Exit Keyboard Binding

Tom-McQuillan committed:
Changed comment colours from yellow to grey

Tom-McQuillan committed:
Non-functional change. Resized functions and set pane origins to 0,0

Tom-McQuillan committed:
Updated whats-new document to state keybinding has been removed, and deleted the RELEASENOTES document as it duplicated efforts

Tom-McQuillan committed:
Merge pull request #39 from JKISoftware/Release-notes

Release notes

Tom-McQuillan committed:
Updated package release description in vipb

jimkring committed:
added new JKI SM Editor package build. incremented vipb build number

jimkring committed:
added new JKI SM Editor package build. incremented vipb build number
